File: gobblet/game.py
"""Gobblet game engine — Python port of game.js + draw rules.

Pure positional logic. The State is an immutable frozen dataclass; apply()
returns a new State (or None for illegal actions). Win detection and the
100-ply cap live in State. 3-fold repetition requires history, so it is
enforced by the Game wrapper used by self-play; MCTS handles cycle detection
along its search path.

Action encoding is size-based (per plan): Action.place(size, to) and
Action.move(from_, to). Piece identity is irrelevant to gameplay (the two
same-size pieces of a color are interchangeable), so the engine never tracks
piece ids.
"""
from __future__ import annotations

import dataclasses
from collections import Counter
from dataclasses import dataclass
from typing import Optional, Tuple

RED, BLUE = 0, 1
S, M, L = 0, 1, 2
SIZES = (S, M, L)
COLORS = (RED, BLUE)
MAX_PLY = 100

LINES = (
    (0, 1, 2), (3, 4, 5), (6, 7, 8),   # rows
    (0, 3, 6), (1, 4, 7), (2, 5, 8),   # cols
    (0, 4, 8), (2, 4, 6),              # diagonals
)
WIN_ORDERS = ((S, M, L), (L, M, S))

CELL = Tuple[int, int]  # (color, size)


def _tray_index(color: int, size: int) -> int:
    return color * 3 + size


@dataclass(frozen=True)
class Action:
    kind: str        # "place" or "move"
    to: int
    size: int = -1   # for place
    from_: int = -1  # for move

    @classmethod
    def place(cls, size: int, to: int) -> "Action":
        return cls(kind="place", to=to, size=size)

    @classmethod
    def move(cls, from_: int, to: int) -> "Action":
        return cls(kind="move", to=to, from_=from_)

    def __repr__(self) -> str:
        if self.kind == "place":
            return f"Place(size={self.size},to={self.to})"
        return f"Move(from={self.from_},to={self.to})"


@dataclass(frozen=True)
class State:
    board: Tuple[Tuple[CELL, ...], ...]   # 9 cells, each a stack bottom->top
    trays: Tuple[int, int, int, int, int, int]  # counts: color*3+size
    player: int                            # player to move
    ply: int
    winner: Optional[int] = None
    is_draw: bool = False

    @classmethod
    def initial(cls) -> "State":
        board = tuple(() for _ in range(9))
        trays = (2, 2, 2, 2, 2, 2)
        return cls(board=board, trays=trays, player=RED, ply=0)

    # ---- accessors ----
    def top_at(self, cell: int) -> Optional[CELL]:
        st = self.board[cell]
        return st[-1] if st else None

    def stack_at(self, cell: int) -> Tuple[CELL, ...]:
        return self.board[cell]

    def tray_count(self, color: int, size: int) -> int:
        return self.trays[_tray_index(color, size)]

    @property
    def current_player(self) -> int:
        return self.player

    def is_terminal(self) -> bool:
        return self.winner is not None or self.is_draw

    def value_for_current(self) -> float:
        if self.winner is not None:
            return 1.0 if self.winner == self.player else -1.0
        if self.is_draw:
            return 0.0
        raise ValueError("state is not terminal")

    @property
    def position_key(self):
        # board + trays + player to move (excludes ply/winner for repetition)
        return (self.board, self.trays, self.player)

    # ---- winning lines ----
    def find_winning_lines(self) -> dict:
        result: dict = {RED: [], BLUE: []}
        for line in LINES:
            tops = []
            ok = True
            for c in line:
                t = self.top_at(c)
                if t is None:
                    ok = False
                    break
                tops.append(t)
            if not ok:
                continue
            color = tops[0][0]
            if any(t[0] != color for t in tops):
                continue
            sizes = tuple(t[1] for t in tops)
            if sizes not in WIN_ORDERS:
                continue
            result[color].append((tuple(line), sizes))
        return result

    def winning_cells(self) -> list:
        if self.winner is None:
            return []
        lines = self.find_winning_lines()
        if lines[self.winner]:
            return list(lines[self.winner][0][0])
        return []

    # ---- legality & apply ----
    def _can_place_on(self, size: int, cell: int) -> bool:
        st = self.board[cell]
        if not st:
            return True
        return size > st[-1][1]

    def legal_actions(self) -> Tuple[Action, ...]:
        if self.is_terminal():
            return ()
        p = self.player
        actions = []
        for sz in SIZES:
            if self.trays[_tray_index(p, sz)] > 0:
                for cell in range(9):
                    if self._can_place_on(sz, cell):
                        actions.append(Action.place(sz, cell))
        for frm in range(9):
            top = self.top_at(frm)
            if top is None or top[0] != p:
                continue
            ps = top[1]
            for to in range(9):
                if to == frm:
                    continue
                if self._can_place_on(ps, to):
                    actions.append(Action.move(frm, to))
        return tuple(actions)

    def apply(self, action: Action) -> Optional["State"]:
        if self.is_terminal():
            return None
        p = self.player
        if action.kind == "place":
            sz, cell = action.size, action.to
            if cell < 0 or cell > 8:
                return None
            if self.trays[_tray_index(p, sz)] <= 0:
                return None
            if not self._can_place_on(sz, cell):
                return None
            new_stack = self.board[cell] + ((p, sz),)
            new_board = self.board[:cell] + (new_stack,) + self.board[cell + 1:]
            new_trays = list(self.trays)
            new_trays[_tray_index(p, sz)] -= 1
            new_trays = tuple(new_trays)
            mover = p
        else:
            frm, to = action.from_, action.to
            if frm < 0 or frm > 8 or to < 0 or to > 8:
                return None
            if frm == to:
                return None
            top = self.top_at(frm)
            if top is None or top[0] != p:
                return None
            if not self._can_place_on(top[1], to):
                return None
            frm_stack = self.board[frm]
            moved = frm_stack[-1]
            new_frm = frm_stack[:-1]
            new_to = self.board[to] + (moved,)
            nb = list(self.board)
            nb[frm] = new_frm
            nb[to] = new_to
            new_board = tuple(nb)
            new_trays = self.trays
            mover = p

        new_ply = self.ply + 1
        new_player = 1 - p
        new_state = State(
            board=new_board, trays=new_trays, player=new_player,
            ply=new_ply, winner=None, is_draw=False,
        )
        wins = new_state.find_winning_lines()
        mover_wins = len(wins[mover]) > 0
        opp_wins = len(wins[1 - mover]) > 0
        if mover_wins and opp_wins:
            new_state = dataclasses.replace(new_state, winner=1 - mover)
        elif mover_wins:
            new_state = dataclasses.replace(new_state, winner=mover)
        elif opp_wins:
            new_state = dataclasses.replace(new_state, winner=1 - mover)
        elif new_ply >= MAX_PLY:
            new_state = dataclasses.replace(new_state, is_draw=True)
        return new_state


class Game:
    """Mutable self-play driver enforcing 3-fold repetition draw on top of State."""

    def __init__(self, state: Optional[State] = None):
        self.state = state if state is not None else State.initial()
        self._counts: Counter = Counter()
        self._counts[self.state.position_key] = 1

    def apply(self, action: Action) -> Optional[State]:
        new_state = self.state.apply(action)
        if new_state is None:
            return None
        if new_state.is_terminal():
            # win or ply-cap already decided; do not override with repetition
            self.state = new_state
            return new_state
        key = new_state.position_key
        self._counts[key] += 1
        if self._counts[key] >= 3:
            new_state = dataclasses.replace(new_state, is_draw=True)
        self.state = new_state
        return new_state

    def is_terminal(self) -> bool:
        return self.state.is_terminal()

    def value_for_current(self) -> float:
        return self.state.value_for_current()

File: ui/game.js
(function (root, factory) {
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = factory();
  } else {
    root.GobbletGame = factory();
  }
})(typeof self !== 'undefined' ? self : this, function () {
  const SIZES = ['S', 'M', 'L'];
  const SIZE_ORDER = { S: 0, M: 1, L: 2 };
  const COLORS = ['red', 'blue'];

  function makePieces(color) {
    const pieces = [];
    let id = 0;
    for (const size of SIZES) {
      for (let i = 0; i < 2; i++) {
        pieces.push({ id: `${color}-${size}-${id++}`, color, size });
      }
    }
    return pieces;
  }

  function create() {
    const state = {
      board: Array.from({ length: 9 }, () => []),
      trays: { red: makePieces('red'), blue: makePieces('blue') },
      currentPlayer: 'red',
      winner: null,
      history: [],
    };

    function other(color) { return color === 'red' ? 'blue' : 'red'; }

    const LINES = [
      [0, 1, 2], [3, 4, 5], [6, 7, 8],   // rows
      [0, 3, 6], [1, 4, 7], [2, 5, 8],   // cols
      [0, 4, 8], [2, 4, 6],              // diagonals
    ];

    function findWinningLines() {
      const result = { red: [], blue: [] };
      for (const line of LINES) {
        const tops = line.map(cell => {
          const stack = state.board[cell];
          return stack.length ? stack[stack.length - 1] : null;
        });
        if (!tops.every(Boolean)) continue;
        const color = tops[0].color;
        if (!tops.every(p => p.color === color)) continue;
        result[color].push({ cells: line.slice(), color });
      }
      return result;
    }

    function snapshot() {
      return {
        board: state.board.map(stack => stack.slice()),
        trays: { red: state.trays.red.slice(), blue: state.trays.blue.slice() },
        currentPlayer: state.currentPlayer,
        winner: state.winner,
      };
    }

    function pushHistory() { state.history.push(snapshot()); }

    function restoreSnapshot(snap) {
      state.board = snap.board.map(s => s.slice());
      state.trays = { red: snap.trays.red.slice(), blue: snap.trays.blue.slice() };
      state.currentPlayer = snap.currentPlayer;
      state.winner = snap.winner;
    }

    function canPlaceOn(piece, cell) {
      const stack = state.board[cell];
      if (!stack.length) return true;
      const top = stack[stack.length - 1];
      return SIZE_ORDER[piece.size] > SIZE_ORDER[top.size];
    }

    function legalActions() {
      if (state.winner) return [];
      const color = state.currentPlayer;
      const actions = [];
      for (const piece of state.trays[color]) {
        for (let cell = 0; cell < 9; cell++) {
          if (canPlaceOn(piece, cell)) actions.push({ type: 'place', pieceId: piece.id, to: cell });
        }
      }
      for (let from = 0; from < 9; from++) {
        const stack = state.board[from];
        if (!stack.length) continue;
        const piece = stack[stack.length - 1];
        if (piece.color !== color) continue;
        for (let to = 0; to < 9; to++) {
          if (to === from) continue;
          if (canPlaceOn(piece, to)) actions.push({ type: 'move', from, to });
        }
      }
      return actions;
    }

    function legalDestinationsForBoardCell(cell) {
      if (state.winner) return [];
      const stack = state.board[cell];
      if (!stack.length) return [];
      const piece = stack[stack.length - 1];
      if (piece.color !== state.currentPlayer) return [];
      const dests = [];
      for (let to = 0; to < 9; to++) {
        if (to === cell) continue;
        if (canPlaceOn(piece, to)) dests.push(to);
      }
      return dests;
    }

    function legalDestinationsForTrayPiece(pieceId) {
      if (state.winner) return [];
      const color = state.currentPlayer;
      const piece = state.trays[color].find(p => p.id === pieceId);
      if (!piece) return [];
      const dests = [];
      for (let to = 0; to < 9; to++) {
        if (canPlaceOn(piece, to)) dests.push(to);
      }
      return dests;
    }

    function applyPlace(pieceId, cell) {
      const color = state.currentPlayer;
      const tray = state.trays[color];
      const idx = tray.findIndex(p => p.id === pieceId);
      if (idx < 0) return { ok: false, reason: 'piece not in tray' };
      const piece = tray[idx];
      const stack = state.board[cell];
      if (stack.length) {
        const top = stack[stack.length - 1];
        if (SIZE_ORDER[piece.size] <= SIZE_ORDER[top.size]) {
          return { ok: false, reason: 'cannot cover same-size or larger piece' };
        }
      }
      pushHistory();
      tray.splice(idx, 1);
      stack.push(piece);
      return { ok: true };
    }

    function applyMove(fromCell, toCell) {
      if (fromCell === toCell) return { ok: false, reason: 'cannot move to the same source cell (no-op)' };
      const stack = state.board[fromCell];
      if (!stack.length) return { ok: false, reason: 'no piece at source cell' };
      const piece = stack[stack.length - 1];
      if (piece.color !== state.currentPlayer) return { ok: false, reason: 'can only move your own top piece' };
      const dest = state.board[toCell];
      if (dest.length) {
        const destTop = dest[dest.length - 1];
        if (SIZE_ORDER[piece.size] <= SIZE_ORDER[destTop.size]) {
          return { ok: false, reason: 'cannot cover same-size or larger piece' };
        }
      }
      pushHistory();
      stack.pop();
      dest.push(piece);
      return { ok: true };
    }

    function resolveAfterAction(mover) {
      const wins = findWinningLines();
      const moverWins = wins[mover].length > 0;
      const opponentWins = wins[other(mover)].length > 0;
      if (moverWins && opponentWins) state.winner = other(mover);
      else if (moverWins) state.winner = mover;
      else if (opponentWins) state.winner = other(mover);
    }

    return {
      get currentPlayer() { return state.currentPlayer; },
      get winner() { return state.winner; },
      isGameOver() { return state.winner !== null; },
      topAt(cell) { const stack = state.board[cell]; return stack.length ? stack[stack.length - 1] : null; },
      stackAt(cell) { return state.board[cell].slice(); },
      tray(color) { return state.trays[color].slice(); },

      findWinningLines() { return findWinningLines(); },

      winningCells() {
        if (!state.winner) return [];
        const lines = findWinningLines();
        const w = state.winner;
        if (lines[w].length) return lines[w][0].cells.slice();
        return [];
      },

      canUndo() { return state.history.length > 0; },

      undo() {
        if (!state.history.length) return { ok: false, reason: 'no history to undo' };
        restoreSnapshot(state.history.pop());
        return { ok: true };
      },

      legalActions() { return legalActions(); },
      legalDestinationsForBoardCell(cell) { return legalDestinationsForBoardCell(cell); },
      legalDestinationsForTrayPiece(pieceId) { return legalDestinationsForTrayPiece(pieceId); },

      serializeState() {
        const COLOR_INT = { red: 0, blue: 1 };
        const SIZE_INT = { S: 0, M: 1, L: 2 };
        const board = state.board.map(stack => stack.map(p => [COLOR_INT[p.color], SIZE_INT[p.size]]));
        const trayCounts = [0, 0, 0, 0, 0, 0];
        for (const color of COLORS) {
          for (const p of state.trays[color]) {
            trayCounts[COLOR_INT[color] * 3 + SIZE_INT[p.size]] += 1;
          }
        }
        return {
          board,
          trays: trayCounts,
          player: COLOR_INT[state.currentPlayer],
          ply: state.history.length,
          winner: state.winner === null ? null : COLOR_INT[state.winner],
          is_draw: false,
        };
      },

      place(pieceId, cell) {
        if (state.winner) return { ok: false, reason: 'game is over' };
        if (cell < 0 || cell > 8) return { ok: false, reason: 'cell out of range' };
        const mover = state.currentPlayer;
        const res = applyPlace(pieceId, cell);
        if (!res.ok) return res;
        resolveAfterAction(mover);
        if (!state.winner) state.currentPlayer = other(mover);
        return { ok: true };
      },

      move(fromCell, toCell) {
        if (state.winner) return { ok: false, reason: 'game is over' };
        if (fromCell < 0 || fromCell > 8) return { ok: false, reason: 'source cell out of range' };
        if (toCell < 0 || toCell > 8) return { ok: false, reason: 'destination cell out of range' };
        const mover = state.currentPlayer;
        const res = applyMove(fromCell, toCell);
        if (!res.ok) return res;
        resolveAfterAction(mover);
        if (!state.winner) state.currentPlayer = other(mover);
        return { ok: true };
      },
    };
  }

  return { create };
});
